Add how to suggest proportion to FAQ (#2718)
optuna/optuna: docs/source/faq.rst

    :class:`~optuna.integration.ChainerMNStudy` does currently not provide ``gc_after_trial`` nor callbacks for :func:`~optuna.integration.ChainerMNStudy.optimize`.
    When using this class, you will have to call the garbage collector inside the objective function.

How do I suggest variables which represent the proportion, that is, are in accordance with Dirichlet distribution?
------------------------------------------------------------------------------------------------------------------

When you want to suggest `n` variables which represent the proportion, that is, `p[0], p[1], ..., p[n-1]` which satisfy `0 <= p[k] <= 1` for any `k` and `p[0] + p[1] + ... + p[n-1] = 1`, try the following.
These variables are in accordance with the Dirichlet distribution.

.. code-block:: python

    import numpy as np
    import matplotlib.pyplot as plt
    import optuna

>     ​

Reviewer: Unnecessary spaces cause the error of the document build:

```suggestion

```

    def objective(trial):
        n = 5
        x = []

>         for i in range(n - 1):

Reviewer: ? 
```suggestion
        for i in range(n:
```

Since https://github.com/optuna/optuna/pull/2718/files#diff-399cec32862278b1cf76d3019a66416c9c5c00e55a067554a69ebebba7684ef3R357 cannot access the last `i`.

    :class:`~optuna.integration.ChainerMNStudy` does currently not provide ``gc_after_trial`` nor callbacks for :func:`~optuna.integration.ChainerMNStudy.optimize`.
    When using this class, you will have to call the garbage collector inside the objective function.

> How do I suggest variables which represent the proportion, that is, are in accordance with Dirichlet distribution?

Reviewer 2: "flat Dirichlet distribution" (or "uniform Dirichlet distribution") is more appropriate.

Author: Thank you for the comment. "flat Dirichlet distribution" sounds more reasonable. I will use it!

------------------------------------------------------------------------------------------------------------------

When you want to suggest `n` variables which represent the proportion, that is, `p[0], p[1], ..., p[n-1]` which satisfy `0 <= p[k] <= 1` for any `k` and `p[0] + p[1] + ... + p[n-1] = 1`, try the following.
These variables are in accordance with the Dirichlet distribution.

.. code-block:: python

    import numpy as np
    import matplotlib.pyplot as plt
    import optuna

    def objective(trial):

>     n = 5
>     x = []
>     for i in range(n):
>         x.append(- np.log(trial.suggest_float(f"x_{i}", 0, 1)))
> 
>         p = []
>         for i in range(n):
>             p.append(x[i] / sum(x))
> 
>         for i in range(n):
>             trial.set_user_attr(f"p_{i}", p[i])
>         return 0

Reviewer: Indents look incorrect according to the example.

```suggestion
        n = 5
        x = []
        for i in range(n):
            x.append(- np.log(trial.suggest_float(f"x_{i}", 0, 1)))
            p = []
        for i in range(n):
            p.append(x[i] / sum(x))
        for i in range(n):
            trial.set_user_attr(f"p_{i}", p[i])
        return 0
```

    study = optuna.create_study(sampler=optuna.samplers.RandomSampler())
    study.optimize(objective, n_trials=1000)

    n = 5
    p = []
    for i in range(n):
        p.append([trial.user_attrs[f"p_{i}"] for trial in study.trials])

>     axes= plt.subplots(n, n, figsize=(20, 20))[1]

Reviewer: ```suggestion
    axes = plt.subplots(n, n, figsize=(20, 20))[1]
```

    for i in range(n):
        for j in range(n):

>             axes[j][i].scatter(p[i], p[j], marker='.')

Reviewer: ```suggestion
            axes[j][i].scatter(p[i], p[j], marker=".")
```

    :class:`~optuna.integration.ChainerMNStudy` does currently not provide ``gc_after_trial`` nor callbacks for :func:`~optuna.integration.ChainerMNStudy.optimize`.
    When using this class, you will have to call the garbage collector inside the objective function.

How do I suggest variables which represent the proportion, that is, are in accordance with Dirichlet distribution?
------------------------------------------------------------------------------------------------------------------

When you want to suggest `n` variables which represent the proportion, that is, `p[0], p[1], ..., p[n-1]` which satisfy `0 <= p[k] <= 1` for any `k` and `p[0] + p[1] + ... + p[n-1] = 1`, try the following.

> These variables are in accordance with the flat Dirichlet distribution.

Reviewer: To make the part more informative, can we add a link to the page of the Dirichlet distribution of Wikipedia https://en.wikipedia.org/wiki/Dirichlet_distribution? I suppose the Dirichlet distribution is not so common for practioners.

    :class:`~optuna.integration.ChainerMNStudy` does currently not provide ``gc_after_trial`` nor callbacks for :func:`~optuna.integration.ChainerMNStudy.optimize`.
    When using this class, you will have to call the garbage collector inside the objective function.

How do I suggest variables which represent the proportion, that is, are in accordance with Dirichlet distribution?
------------------------------------------------------------------------------------------------------------------

> When you want to suggest :math:`n` variables which represent the proportion, that is, :math:`p[0], p[1], ..., p[n-1]` which satisfy :math:`0 \le p[k] \le 1` for any :math:`k` and :math:`p[0] + p[1] + ... + p[n-1] = 1`, try the following.

Reviewer 3: ```suggestion
When you want to suggest :math:`n` variables which represent the proportion, that is, :math:`p[0], p[1], ..., p[n-1]` which satisfy :math:`0 \le p[k] \le 1` for any :math:`k` and :math:`p[0] + p[1] + ... + p[n-1] = 1`, try the below.
```

These variables are in accordance with the flat `Dirichlet distribution <https://en.wikipedia.org/wiki/Dirichlet_distribution>`_.

.. code-block:: python

    import numpy as np
    import matplotlib.pyplot as plt
    import optuna

    def objective(trial):
        n = 5
        x = []
        for i in range(n):
            x.append(- np.log(trial.suggest_float(f"x_{i}", 0, 1)))

        p = []
        for i in range(n):
            p.append(x[i] / sum(x))

        for i in range(n):
            trial.set_user_attr(f"p_{i}", p[i])

        return 0

    study = optuna.create_study(sampler=optuna.samplers.RandomSampler())
    study.optimize(objective, n_trials=1000)

    n = 5
    p = []
    for i in range(n):
        p.append([trial.user_attrs[f"p_{i}"] for trial in study.trials])
    axes = plt.subplots(n, n, figsize=(20, 20))[1]

    for i in range(n):
        for j in range(n):
            axes[j][i].scatter(p[i], p[j], marker=".")
            axes[j][i].set_xlim(0, 1)
            axes[j][i].set_ylim(0, 1)
            axes[j][i].set_xlabel(f"p_{i}")
            axes[j][i].set_ylabel(f"p_{j}")

    plt.savefig("sampled_ps.png")

> This method is justified in the following way.

Reviewer 3: ```suggestion
This method is justified in the following way:
```

First, if we apply the transformation :math:`x = - \log (u)` to the variable :math:`u` sampled from the uniform distribution :math:`Uni(0, 1)` in the interval :math:`[0, 1]`, the variable :math:`x` will follow the exponential distribution :math:`Exp(1)` with scale parameter :math:`1`.
Furthermore, for :math:`n` variables :math:`x[0], ..., x[n-1]` that follow the exponential distribution of scale parameter :math:`1` independently, normalizing them with :math:`p[i] = x[i] / \sum_i x[i]`, the vector :math:`p` follows the Dirichlet distribution :math:`Dir(\alpha)` of scale parameter :math:`\alpha = (1, ..., 1)`.

> You can verify the transformation by elemental calculation of Jacobian.

Reviewer 3: ```suggestion
You can verify the transformation by calculating the elements of the Jacobian.
```
